### PR TITLE
fix(react): Support React 18 types

### PR DESCRIPTION
**Pull request by stephenmathieson**
This patch updates `@deque/cauldron-react` to play nicely with [React v18 types](https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html#updates-to-typescript-definitions).

Part of https://github.com/dequelabs/walnut/issues/3349

### CODE REVIEW

File: packages/react/src/components/Tooltip/index.tsx [hunk 1]
```diff
@@ -178,34 +178,43 @@ export default function Tooltip({
...
-          })}
-          ref={setTooltipElement}
-          role="tooltip"
-          style={styles.popper}
-          {...attributes.popper}
-          {...props}
-        >
-          {variant !== 'big' && (
+  return (
+    <>
```
Comment (scurker) [inline packages/react/src/components/Tooltip/index.tsx:182]: Why the change to fragment here?

File: packages/react/src/components/TwoColumnPanel/TwoColumnPanel.tsx [hunk 2]
```diff
@@ -231,22 +232,24 @@ const TwoColumnPanel = forwardRef<HTMLDivElement, TwoColumnPanelProps>(
...
-          containerElements={[columnLeftRef.current as HTMLElement]}
-        />
-        <ClickOutsideListener
-          onClickOutside={handleClickOutside}
-          target={columnLeftRef.current as HTMLElement}
-        />
-        {isCollapsed ? null : skipLink}
-        {showPanel ? ColumnLeftComponent : null}
-        {ColumnRightComponent}
+        <>
```
Comment (scurker) [inline packages/react/src/components/TwoColumnPanel/TwoColumnPanel.tsx:235]: ?

File: packages/react/src/utils/polymorphic-type/index.ts [hunk 3]
```diff
@@ -51,7 +51,7 @@ interface ForwardRefComponent<
   <As = IntrinsicElementString>(
     props: As extends ''
       ? { as: keyof JSX.IntrinsicElements }
-      : As extends React.ComponentType<infer P>
+      : As extends React.ComponentType<React.PropsWithChildren<infer P>>
```
Comment (scurker) [inline packages/react/src/utils/polymorphic-type/index.ts:54]: Is this a safe assumption to make? I don't know that the polymorphic type will always be on a component with children.

File: packages/react/src/components/TwoColumnPanel/TwoColumnPanel.tsx (hunk 2 above)
Comment (stephenmathieson) [reply packages/react/src/components/TwoColumnPanel/TwoColumnPanel.tsx:235]: I no longer have the exact error message, but it was something like...

> `<div>` is expecting a single child, but multiple were provided

This was the easiest/less intrusive fix, as it has 0 effect on anything other than TypeScript.

File: packages/react/src/utils/polymorphic-type/index.ts (hunk 3 above)
Comment (stephenmathieson) [reply packages/react/src/utils/polymorphic-type/index.ts:54]: This was done by the codemod (not by hand). I'll see what happens when I revert this change.

File: packages/react/src/components/Tooltip/index.tsx (hunk 1 above)
Comment (scurker) [reply packages/react/src/components/Tooltip/index.tsx:182]: Interesting, I thought React 18 allowed returning `null` but 🤷

**Review by scurker — APPROVED** [review]
